realm/realm-library/src/main/java/io/realm/internal/Table.java

```diff
@@ -196,14 +198,35 @@ public long addColumnLink (RealmFieldType type, String name, Table table) {
     @Override
     public void removeColumn(long columnIndex) {
         nativeRemoveColumn(nativePtr, columnIndex);
+        // When PK does not exist this cannot happen. But if you remove a column with a smaller index
+        // than that of PK column, you need to rearrange PK column index.
+        if (hasPrimaryKey() && columnIndex <= getPrimaryKey()) {
+            invalidateCachedPrimaryKeyIndex();
+        }
     }
 
     /**
-     * Renames a column in the table.
+     * Renames a column in the table. This needs to be called manually before inserting data into the table.
```

Comment:
Not sure I understand "This needs to be called manually before inserting data into the table." ?